### PR TITLE
Change generate-document Lambda to conditionally generate a Word format DoW document

### PR DESCRIPTION
**Pull request by zachclark-ccpo**
## Overview
In-progress draft PR...

- [x] Tables that scale with array size for Award History table at the top of DOW
- [x] Mapping Nested Objects to Document
- [x] Bullet points
- [x] API Request returns buffer correctly

## Handling Nested objects
image

Input:
```
const test_body = {
    documentType: "DESCRIPTION_OF_WORK",
    templatePayload: {
        contractInformation: {
            contractNumber: "1234567890",
            currentContractExists: true,
            contractExpirationDate: "2022-09-30",
            incumbentContractorName: "Morgan Hall",
            previousTaskOrderNumber: "0987654321",
        },
```
Output:
image

## Handling Tables on the template
Input:
image

Taking the payload as the `templatePayload` from 
```typescript
export interface GenerateDocumentRequest {
  documentType: DocumentType;
  templatePayload: DescriptionOfWork | IndependentGovernmentCostEstimate;
}
```
`const test_payload = test_body.templatePayload;`

Below is a small snippet featuring the `awardHistory`, but the intention is to pass the entire `templatePayload` to `document.render(payload)`
```typescript 
{
awardHistory: [
    {
        contractAwardType: "INITIAL_AWARD",
        effectiveDate: "2022-05-20",
        modificationOrder: "",
    },
    {
        contractAwardType: "MODIFICATION",
        modificationOrder: 1,
        effectiveDate: "2022-09-01",
    },
    {
        contractAwardType: "MODIFICATION",
        modificationOrder: 3,
        effectiveDate: "2023-02-01",
    },
],    
},
```  

Output:
image

Documentation Referenced:
https://docxtemplater.com/docs/tag-types/#creating-multiple-table-rows
https://docxtemplater.com/docs/tag-types/#loops

### CODE REVIEW

File: document-generation/.docx/generateDow.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import * as path from "path";
+import PizZip from "pizzip";
+import Docxtemplater from "docxtemplater";
+
+export async function generateDow(event: RequestEvent<GenerateDocumentRequest>): Promise<ApiBase64SuccessResponse> {
+  // Extract payload
+  const payload = event.body.templatePayload;
+
+  // Load the docx file as binary content
+  const content = fs.readFileSync(path.resolve(__dirname, "dow.docx"), "binary");
```
Comment (juliusfitzhugh-ccpo) [inline document-generation/.docx/generateDow.ts:13]: The `__dirname` property inside of lambda for this file is `/var/task` so the final path ends up being `/var/task/dow.docx`, but the template is stored in `/opt/dow.docx` due to a layer being used for the templates (see [dependencies in a layer](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#configuration-layers-path)). 

### Suggestions:
- remove `path.resolve()` and simply provided `/opt/dow.docx` for file path
- use the `getDocxTemplatePath()` in the `utils/utils.ts` to return the file path
- use `path` to create the correct path

**Review by juliusfitzhugh-ccpo — CHANGES_REQUESTED** [review]
Overall looks good, I like the addition of the folder structure with `.docx` I think there is room for refactoring/organizing the generated documents based on file types (not now, but in the near future).  

Two minor changes in the `generateDow.ts` file is needed to return the generated Dow document from the Lambda and API Gateway.

File: document-generation/generate-document.ts [hunk 2]
```diff
@@ -27,13 +27,18 @@ import {
...
 async function baseHandler(event: RequestEvent<GenerateDocumentRequest>): Promise<ApiBase64SuccessResponse> {
   const { documentType } = event.body;
   logger.info("Generating document", { documentType });
 
   switch (documentType) {
-    case DocumentType.DESCRIPTION_OF_WORK:
+    // case document type = document type enum
+    // return generateDOCUMENTNAME (located in docx folder)
+    case DocumentType.DESCRIPTION_OF_WORK_DOCX:
+      return generateDow(event);
```
[the remark this answers is not in the document]
Comment (zachclark-ccpo) [reply document-generation/generate-document.ts:40]: Good catch, I'll clean up those comments.

File: document-generation/.docx/generateDow.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const document = new Docxtemplater(zip, {
+    paragraphLoop: true,
+    linebreaks: true,
+  });
+
+  document.render(payload);
+
+  // Create Node JS Buffer to send as response
+  const report = document.getZip().generate({
+    type: "nodebuffer",
```
[the remark this answers is not in the document]
Comment (zachclark-ccpo) [reply document-generation/.docx/generateDow.ts:26]: After experimenting, went with:  
```
// Create Node JS Buffer to send as response
  const report = Buffer.from(document.getZip().generate({
    type: "nodebuffer",
    compression: "DEFLATE",
  }));
```
The `base64` string default resulted in a corrupted word document.

[the remark this answers is not in the document]
Comment (juliusfitzhugh-ccpo) [reply document-generation/.docx/generateDow.ts:26]: Hmm that's weird...well if it works that way 🤷

**Review by juliusfitzhugh-ccpo — APPROVED** [review]
Pulled down and ran branch on a deployed stack and was able to generate the DOW document. 

Besides the secrets baseline and test fixes, should be good to go.

**Pull request closed without merging**